File: fries-leetcode-training/09-backtracking/02-permutations-ii.cpp
/**
 * LeetCode 47. 全排列II
 * https://leetcode.cn/problems/permutations-ii/
 * 
 * 给定一个可包含重复数字的序列nums，按任意顺序返回所有不重复的全排列。
 * 
 * 回溯 + 去重
 * 
 * 时间复杂度：O(n! * n)
 * 空间复杂度：O(n)
 */

class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> result;
        vector<int> path;
        vector<bool> used(nums.size(), false);
        
        sort(nums.begin(), nums.end());  // 排序以便去重
        
        backtrack(nums, path, used, result);
        return result;
    }
    
private:
    void backtrack(vector<int>& nums, vector<int>& path, vector<bool>& used, vector<vector<int>>& result) {
        if (path.size() == nums.size()) {
            result.push_back(path);
            return;
        }
        
        for (int i = 0; i < nums.size(); i++) {
            if (used[i]) {
                continue;
            }
            
            // 去重：如果当前数字与前一个数字相同，且前一个数字未被使用，则跳过
            if (i > 0 && nums[i] == nums[i-1] && !used[i-1]) {
                continue;
            }
            
            path.push_back(nums[i]);
            used[i] = true;
            backtrack(nums, path, used, result);
            path.pop_back();
            used[i] = false;
        }
    }
};

// 测试函数
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void testPermuteUnique() {
    Solution solution;
    
    vector<int> nums = {1, 1, 2};
    vector<vector<int>> result = solution.permuteUnique(nums);
    
    cout << "全排列结果: [";
    for (int i = 0; i < result.size(); i++) {
        cout << "[";
        for (int j = 0; j < result[i].size(); j++) {
            cout << result[i][j];
            if (j < result[i].size() - 1) cout << ",";
        }
        cout << "]";
        if (i < result.size() - 1) cout << ",";
    }
    cout << "]" << endl;  // 期望: [[1,1,2],[1,2,1],[2,1,1]]
}

int main() {
    testPermuteUnique();
    return 0;
}

File: fries-leetcode-training/09-backtracking/02-permutations-ii.py
"""
LeetCode 47. 全排列II
https://leetcode.cn/problems/permutations-ii/

给定一个可包含重复数字的序列nums，按任意顺序返回所有不重复的全排列。

回溯 + 去重

时间复杂度：O(n! * n)
空间复杂度：O(n)
"""

def permute_unique(nums):
    """
    全排列II - 回溯 + 去重法
    
    Args:
        nums: 可包含重复数字的序列
        
    Returns:
        所有不重复的全排列
    """
    result = []
    nums.sort()  # 排序以便去重
    
    def backtrack(path, used):
        if len(path) == len(nums):
            result.append(path[:])
            return
        
        for i in range(len(nums)):
            if used[i]:
                continue
            
            # 去重：如果当前数字与前一个数字相同，且前一个数字未被使用，则跳过
            if i > 0 and nums[i] == nums[i-1] and not used[i-1]:
                continue
            
            path.append(nums[i])
            used[i] = True
            backtrack(path, used)
            path.pop()
            used[i] = False
    
    backtrack([], [False] * len(nums))
    return result


def test_permute_unique():
    """测试函数"""
    # 测试用例1
    nums1 = [1, 1, 2]
    result1 = permute_unique(nums1)
    print(f"测试1 [1,1,2]: {result1}")
    # 期望: [[1,1,2],[1,2,1],[2,1,1]]
    
    # 测试用例2
    nums2 = [1, 2, 3]
    result2 = permute_unique(nums2)
    print(f"测试2 [1,2,3]: {result2}")
    # 期望: [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]


if __name__ == "__main__":
    test_permute_unique()
